functions/access.js

```diff
@@ -0,0 +1,136 @@
+      }
+
+      return false;
+    })
+    .catch((e) => {
+      console.log("Transaction failure:", e);
+      return false;
+    });
+};
+
+const getAccessDoc = async (venueId, method) => {
+  const venue = await admin.firestore().collection("venues").doc(venueId).get();
+  if (!venue.exists) {
+    throw new HttpsError("not-found", `venue ${venueId} does not exist`);
+  }
+  const accessDoc = await venue.ref.collection("access").doc(method).get();
+  return accessDoc;
+};
+
+const isValidPassword = async (venueId, password) => {
+  const access = await getAccessDoc(venueId, "password");
+  if (!access.exists || !access.data().password) {
+    return false;
+  }
+  return passwordsMatch(access.data().password, password);
+};
+
+const isValidEmail = async (venueId, email) => {
+  const access = await getAccessDoc(venueId, "emails");
+  if (!access.exists || !access.data().emails) {
+    return false;
+  }
+  return access.data().emails.includes(email.trim().toLowerCase());
+};
+
+const isValidCode = async (venueId, code) => {
+  const access = getAccessDoc(venueId, "code");
+  if (!access.exists || !access.data().codes) {
+    return false;
+  }
+  return access.data().codes.includes(code.trim());
+};
+
+const createToken = async (venueId, uid, password, email, code) => {
+  if (!venueId || !uid || !password || !email || !code) return {};
+
+  const venue = await admin.firestore().collection("venues").doc(venueId).get();
+  if (!venue.exists) {
+    throw new HttpsError("not-found", `venue ${venueId} does not exist`);
+  }
+
+  const token = uuid();
+  const grantedDoc = await venue.ref.collection("accessgranted").doc(uid).get();
+
+  const tokenData = {
+    usedAt: [Date.now()],
+    password: password || null,
+    email: email || null,
+    code: code || null,
+  };
+
+  if (grantedDoc.exists) {
+    await grantedDoc.update({ [`tokens.${token}`]: tokenData });
+  } else {
+    const granted = { tokens: { [token]: tokenData } };
+    await grantedDoc.set(granted);
+  }
+  return { token };
+};
+
+exports.checkAccess = functions.https.onCall(async (data, context) => {
+  console.log("checkAccess", data, context.auth);
+  const isValidToken =
+    context &&
+    context.auth &&
+    context.auth.uid &&
+    (await checkIsValidToken(data.venueId, context.auth.uid, data.token));
+  if (isValidToken) {
+    console.log(`valid token, returning ${data.token}`);
+    return { token: data.token };
+  }
+
+  const isPasswordValid =
+    data.password && (await isValidPassword(data.venueId, data.password));
+  const isEmailValid =
+    data.email && (await isValidEmail(data.venueId, data.email));
+  const isCodeValid = data.code && (await isValidCode(data.venueId, data.code));
+
+  if (isPasswordValid || isEmailValid || isCodeValid) {
+    const token = await createToken(
+      data.venueId,
+      context.auth.uid,
+      data.password,
+      data.email,
+      data.code
+    );
+    console.log(`create token: ${token}`);
+    return { token };
```

Comment:
Since `createToken` currently returns `{ token }` (prior to suggestions in this review being implemented), then I believe this is going to effectively end up returning `{ { token } }`.

I'm guessing that `functions.https.onCall()` probably needs us to return an object here (that I assume it will convert to a JSON return value/similar)? If so, then just fixing `createToken` to not return an object wrapper should be sufficient.